### PR TITLE
feat(manager/poetry): extract python as a dependency from `pyproject.toml`

### PR DESCRIPTION
**Pull request by zeshuaro**
## Changes

- Copied from last PR: https://github.com/renovatebot/renovate/pull/24236
  - Update poetry manager to extract python as a dependency from `pyproject.toml` so that its version can be updated
  - Update `artifact.getPythonConstraint` to first attempt to read the python version from `pyproject.toml` first as it could have been updated then fallback to read from `poetry.lock`
- **New changes** to explicitly set `registryUrls` to null for the extracted python dependency due to this reported issue
  - https://github.com/renovatebot/renovate/discussions/24256

## Context

- https://github.com/renovatebot/renovate/discussions/19144

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository
  - https://github.com/barriot/r0/pull/16 - updated python version
  - https://github.com/barriot/r0/pull/11 - confirmed that existing package update still works

### CODE REVIEW

**Review by rarkins — COMMENTED** [review]
I would like to see this tested against the type of repo which experienced the problem previously, to make sure the registry URLs problem is definitely gone

**Comment by zeshuaro** [comment]
I ran it through the test repo with the update `pyproject.toml` to include a source/registry URL config: https://github.com/barriot/r0/blob/main/pyproject.toml

And here's the logs where there's no warnings:

<details>
  <summary>Logs</summary>

  ```logging
  DEBUG: Using RE2 as regex engine
DEBUG: Parsing configs
DEBUG: Checking for config file in <redacted-config-path>
DEBUG: No config file found on disk - skipping
DEBUG: File config
       "config": {}
DEBUG: CLI config
       "config": {
         "repositories": ["barriot/r0"],
         "binarySource": "docker",
         "enabledManagers": ["poetry"]
       }
DEBUG: Env config
       "config": {"hostRules": [], "token": "***********"}
DEBUG: Combined config
       "config": {
         "hostRules": [],
         "token": "***********",
         "repositories": ["barriot/r0"],
         "binarySource": "docker",
         "enabledManagers": ["poetry"]
       }
DEBUG: Enabling forkProcessing while in non-autodiscover mode
DEBUG: Found valid git version: 2.38.1
DEBUG: Using default github endpoint: https://api.github.com/
DEBUG: GET https://api.github.com/user/emails = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=353)
DEBUG: Cannot read user/emails endpoint on GitHub to retrieve gitAuthor
DEBUG: Platform config
       "platformConfig": {
         "hostType": "github",
         "endpoint": "https://api.github.com/",
         "isGHApp": false,
         "isGhe": false,
         "userDetails": {"username": "barriot", "name": null, "id": 21984486},
         "userEmail": null
       },
       "renovateUsername": "barriot"
DEBUG: Adding token authentication for api.github.com to hostRules
DEBUG: Using baseDir: /var/folders/dj/wt5h_b751qjfgf5h0fv5bkvh0000gn/T/renovate
DEBUG: Using cacheDir: /var/folders/dj/wt5h_b751qjfgf5h0fv5bkvh0000gn/T/renovate/cache
DEBUG: Using containerbaseDir: /var/folders/dj/wt5h_b751qjfgf5h0fv5bkvh0000gn/T/renovate/cache/containerbase
DEBUG: Initializing Renovate internal cache into /var/folders/dj/wt5h_b751qjfgf5h0fv5bkvh0000gn/T/renovate/cache/renovate/renovate-cache-v1
DEBUG: Commits limit = null
DEBUG: Setting global hostRules
DEBUG: Adding token authentication for api.github.com to hostRules
DEBUG: validatePresets()
DEBUG: Reinitializing hostRules for repo
DEBUG: Clearing hostRules
DEBUG: Adding token authentication for api.github.com to hostRules
DEBUG: No dangling containers to remove
 INFO: Repository started (repository=barriot/r0)
       "renovateVersion": "0.0.0-semantic-release"
DEBUG: Using localDir: /var/folders/dj/wt5h_b751qjfgf5h0fv5bkvh0000gn/T/renovate/repos/github/barriot/r0 (repository=barriot/r0)
DEBUG: PackageFiles.clear() - Package files deleted (repository=barriot/r0)
DEBUG: initRepo("barriot/r0") (repository=barriot/r0)
DEBUG: barriot/r0 default branch = main (repository=barriot/r0)
DEBUG: Using personal access token for git init (repository=barriot/r0)
DEBUG: Resetting npmrc (repository=barriot/r0)
DEBUG: Resetting npmrc (repository=barriot/r0)
DEBUG: checkOnboarding() (repository=barriot/r0)
DEBUG: isOnboarded() (repository=barriot/r0)
DEBUG: findPr(renovate/configure, Configure Renovate, !open) (repository=barriot/r0)
DEBUG: getPrList success (repository=barriot/r0)
       "pullsTotal": 15,
       "requestsTotal": 1,
       "apiQuotaAffected": true
DEBUG: Found PR #1 (repository=barriot/r0)
DEBUG: findFile(renovate.json) (repository=barriot/r0)
DEBUG: Initializing git repository into /var/folders/dj/wt5h_b751qjfgf5h0fv5bkvh0000gn/T/renovate/repos/github/barriot/r0 (repository=barriot/r0)
DEBUG: Performing blobless clone (repository=barriot/r0)
DEBUG: git clone completed (repository=barriot/r0)
       "durationMs": 2780
DEBUG: latest repository commit (repository=barriot/r0)
       "latestCommit": {
         "hash": "a864255b0b4159f2592027b1cd2dd6cc78bc5163",
         "date": "2023-09-10T10:37:12+10:00",
         "message": "test",
         "refs": "HEAD -> main, origin/main, origin/HEAD",
         "body": "",
         "author_name": "<redacted-name>",
         "author_email": "<redacted-email>"
       }
DEBUG: Config file exists, fileName: renovate.json (repository=barriot/r0)
DEBUG: Retrieving issueList (repository=barriot/r0)
DEBUG: Retrieved 1 issues (repository=barriot/r0)
DEBUG: Repo is onboarded (repository=barriot/r0)
DEBUG: Found renovate.json config file (repository=barriot/r0)
DEBUG: Repository config (repository=barriot/r0)
       "fileName": "renovate.json",
       "config": {
         "$schema": "https://docs.renovatebot.com/renovate-schema.json",
         "ignorePaths": ["android/**", "ios/**", "macos/**"],
         "rangeStrategy": "bump"
       }
DEBUG: migrateAndValidate() (repository=barriot/r0)
DEBUG: No config migration necessary (repository=barriot/r0)
DEBUG: Found repo ignorePaths (repository=barriot/r0)
       "ignorePaths": ["android/**", "ios/**", "macos/**"]
DEBUG: No vulnerability alerts found (repository=barriot/r0)
DEBUG: No vulnerability alerts found (repository=barriot/r0)
DEBUG: No baseBranches (repository=barriot/r0)
DEBUG: extract() (repository=barriot/r0)
DEBUG: Setting current branch to main (repository=barriot/r0)
DEBUG: latest commit (repository=barriot/r0)
       "branchName": "main",
       "latestCommitDate": "2023-09-10T10:37:12+10:00"
DEBUG: Applying enabledManagers filtering (repository=barriot/r0)
DEBUG: Using file match: (^|/)pyproject\.toml$ for manager poetry (repository=barriot/r0)
DEBUG: Matched 1 file(s) for manager poetry: pyproject.toml (repository=barriot/r0)
DEBUG: manager extract durations (ms) (repository=barriot/r0)
       "managers": {"poetry": 8}
DEBUG: Found poetry package files (repository=barriot/r0)
DEBUG: Found 1 package file(s) (repository=barriot/r0)
 INFO: Dependency extraction complete (repository=barriot/r0, baseBranch=main)
       "stats": {
         "managers": {"poetry": {"fileCount": 1, "depCount": 2}},
         "total": {"fileCount": 1, "depCount": 2}
       }
DEBUG: GET https://foo.bar/simple/requests/ = (code=ENOTFOUND, statusCode=-1 retryCount=2, duration=2) (repository=barriot/r0)
DEBUG: PackageFiles.add() - Package file saved for base branch (repository=barriot/r0, baseBranch=main)
DEBUG: Package releases lookups complete (repository=barriot/r0, baseBranch=main)
DEBUG: branchifyUpgrades (repository=barriot/r0)
DEBUG: detectSemanticCommits() (repository=barriot/r0)
DEBUG: getCommitMessages (repository=barriot/r0)
DEBUG: semanticCommits: detected "unknown" (repository=barriot/r0)
DEBUG: semanticCommits: disabled (repository=barriot/r0)
DEBUG: 2 flattened updates found: python, requests (repository=barriot/r0)
DEBUG: Returning 2 branch(es) (repository=barriot/r0)
DEBUG: config.repoIsOnboarded=true (repository=barriot/r0)
DEBUG: packageFiles with updates (repository=barriot/r0, baseBranch=main)
       "config": {
         "poetry": [
           {
             "deps": [
               {
                 "datasource": "docker",
                 "currentValue": "3.10.11",
                 "managerData": {"nestedVersion": false},
                 "versioning": "pep440",
                 "depName": "python",
                 "depType": "dependencies",
                 "commitMessageTopic": "Python",
                 "registryUrls": null,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "3.11.5",
                     "newValue": "3.11.5",
                     "newMajor": 3,
                     "newMinor": 11,
                     "updateType": "minor",
                     "branchName": "renovate/python-3.x"
                   }
                 ],
                 "packageName": "python",
                 "warnings": [],
                 "registryUrl": "https://index.docker.io",
                 "currentVersion": "3.10.11",
                 "isSingleVersion": true,
                 "fixedVersion": "3.10.11"
               },
               {
                 "datasource": "pypi",
                 "currentValue": "^2.27.0",
                 "managerData": {"nestedVersion": false},
                 "versioning": "poetry",
                 "depName": "requests",
                 "depType": "dependencies",
                 "lockedVersion": "2.27.0",
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "2.31.0",
                     "newValue": "^2.31.0",
                     "releaseTimestamp": "2023-05-22T15:12:42.000Z",
                     "newMajor": 2,
                     "newMinor": 31,
                     "updateType": "minor",
                     "isRange": true,
                     "isBump": true,
                     "branchName": "renovate/requests-2.x"
                   }
                 ],
                 "packageName": "requests",
                 "warnings": [],
                 "sourceUrl": "https://github.com/psf/requests",
                 "registryUrl": "https://pypi.org/pypi",
                 "homepage": "https://requests.readthedocs.io",
                 "changelogUrl": "https://github.com/psf/requests/blob/master/HISTORY.md",
                 "currentVersion": "2.27.0",
                 "isSingleVersion": false,
                 "fixedVersion": "2.27.0"
               }
             ],
             "registryUrls": ["https://foo.bar/simple/", "https://pypi.org/pypi/"],
             "extractedConstraints": {"python": "3.10.11"},
             "lockFiles": ["poetry.lock"],
             "packageFile": "pyproject.toml"
           }
         ]
       }
DEBUG: detectSemanticCommits() (repository=barriot/r0)
DEBUG: semanticCommits: returning "disabled" from cache (repository=barriot/r0)
DEBUG: processRepo() (repository=barriot/r0)
DEBUG: Processing 2 branches: renovate/python-3.x, renovate/requests-2.x (repository=barriot/r0)
DEBUG: Calculating hourly PRs remaining (repository=barriot/r0)
DEBUG: currentHourStart=2023-09-10T11:00:00.000+10:00 (repository=barriot/r0)
DEBUG: PR hourly limit remaining: 2 (repository=barriot/r0)
DEBUG: Calculating prConcurrentLimit (10) (repository=barriot/r0)
DEBUG: getBranchPr(renovate/requests-2.x) (repository=barriot/r0)
DEBUG: findPr(renovate/requests-2.x, undefined, open) (repository=barriot/r0)
DEBUG: Found PR #11 (repository=barriot/r0)
DEBUG: getBranchPr(renovate/python-3.x) (repository=barriot/r0)
DEBUG: findPr(renovate/python-3.x, undefined, open) (repository=barriot/r0)
DEBUG: Found PR #16 (repository=barriot/r0)
DEBUG: 2 PRs are currently open (repository=barriot/r0)
DEBUG: PR concurrent limit remaining: 8 (repository=barriot/r0)
DEBUG: Calculated maximum PRs remaining this run: 2 (repository=barriot/r0)
DEBUG: PullRequests limit = 2 (repository=barriot/r0)
DEBUG: Calculating hourly PRs remaining (repository=barriot/r0)
DEBUG: currentHourStart=2023-09-10T11:00:00.000+10:00 (repository=barriot/r0)
DEBUG: PR hourly limit remaining: 2 (repository=barriot/r0)
DEBUG: Calculating branchConcurrentLimit (10) (repository=barriot/r0)
DEBUG: 2 already existing branches found: renovate/requests-2.x,renovate/python-3.x (repository=barriot/r0)
DEBUG: Branch concurrent limit remaining: 8 (repository=barriot/r0)
DEBUG: Calculated maximum branches remaining this run: 2 (repository=barriot/r0)
DEBUG: Branches limit = 2 (repository=barriot/r0)
DEBUG: syncBranchState() (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: syncBranchState(): Branch cache not found, creating minimal branchState (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: getBranchPr(renovate/requests-2.x) (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: findPr(renovate/requests-2.x, undefined, open) (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Found PR #11 (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: branchExists=true (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: dependencyDashboardCheck=undefined (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: PR rebase requested=false (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Checking if PR has been edited (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: branch.isModified(): using git to calculate (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: branch.isModified() = false (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Found existing branch PR (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Checking schedule(at any time, null) (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: No schedule defined (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Branch already exists (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: GET https://api.github.com/repos/barriot/r0/branches/main/protection = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=623) (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: No branch protection found (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Skipping behind base branch check due to rebaseWhen=auto (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: isBranchConflicted(main, renovate/requests-2.x) (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: branch.isConflicted(): using git to calculate (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Setting git author name: Renovate Bot (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Setting git author email: renovate@whitesourcesoftware.com (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: branch.isConflicted(): false (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Branch does not need rebasing (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Using reuseExistingBranch: true (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Setting current branch to main (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: latest commit (repository=barriot/r0, branch=renovate/requests-2.x)
       "branchName": "main",
       "latestCommitDate": "2023-09-10T10:37:12+10:00"
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=true (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Branch dep requests in pyproject.toml is already updated (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: No content changed (repository=barriot/r0, packageFile=pyproject.toml, branch=renovate/requests-2.x)
       "depName": "requests"
DEBUG: No package files need updating (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: No updated lock files in branch (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: No files to commit (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Setting current branch to main (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: latest commit (repository=barriot/r0, branch=renovate/requests-2.x)
       "branchName": "main",
       "latestCommitDate": "2023-09-10T10:37:12+10:00"
DEBUG: Checking if we can automerge branch (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: mergeStatus=no automerge (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Ensuring PR (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: There are 0 errors and 0 warnings (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: getBranchPr(renovate/requests-2.x) (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: findPr(renovate/requests-2.x, undefined, open) (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Found PR #11 (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: getPrCache() (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Found existing PR (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Fetching changelog: https://github.com/psf/requests (2.27.0 -> 2.31.0) (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Processing existing PR (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: setPrCache() (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: Pull Request #11 does not need updating (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: PR is not configured for automerge (repository=barriot/r0, branch=renovate/requests-2.x)
DEBUG: syncBranchState() (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: syncBranchState(): Branch cache not found, creating minimal branchState (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: getBranchPr(renovate/python-3.x) (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: findPr(renovate/python-3.x, undefined, open) (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Found PR #16 (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: branchExists=true (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: dependencyDashboardCheck=undefined (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Manual rebase requested via PR checkbox for #16 (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: PR rebase requested=true (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Checking if PR has been edited (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: branch.isModified(): using git to calculate (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: branch.isModified() = false (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Found existing branch PR (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Checking schedule(at any time, null) (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: No schedule defined (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: User has requested rebase (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Using reuseExistingBranch: false (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Setting current branch to main (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: latest commit (repository=barriot/r0, branch=renovate/python-3.x)
       "branchName": "main",
       "latestCommitDate": "2023-09-10T10:37:12+10:00"
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=false (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Starting search at index 114 (repository=barriot/r0, packageFile=pyproject.toml, branch=renovate/python-3.x)
       "depName": "python"
DEBUG: Found match at index 114 (repository=barriot/r0, packageFile=pyproject.toml, branch=renovate/python-3.x)
       "depName": "python"
DEBUG: Contents updated (repository=barriot/r0, packageFile=pyproject.toml, branch=renovate/python-3.x)
       "depName": "python"
DEBUG: poetry.updateArtifacts(pyproject.toml) (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Updating poetry.lock (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Using python version from pyproject.toml (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Using poetry version from poetry.lock header (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /var/folders/dj/wt5h_b751qjfgf5h0fv5bkvh0000gn/T/renovate/cache/containerbase (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Using docker to execute (repository=barriot/r0, branch=renovate/python-3.x)
       "image": "sidecar"
DEBUG: containerbaseDir is inside cacheDir (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Resolved tag constraint (repository=barriot/r0, branch=renovate/python-3.x)
       "image": "ghcr.io/containerbase/sidecar:9.19.5"
DEBUG: Fetching Docker image: ghcr.io/containerbase/sidecar:9.19.5 (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Finished fetching Docker image ghcr.io/containerbase/sidecar:9.19.5@sha256:6aa7f43974ec96ede2b783103ac615ca1ce89ea023bdbf30d22aa3d23d67b127 (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Executing command (repository=barriot/r0, branch=renovate/python-3.x)
       "command": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/var/folders/dj/wt5h_b751qjfgf5h0fv5bkvh0000gn/T/renovate/repos/github/barriot/r0\":\"/var/folders/dj/wt5h_b751qjfgf5h0fv5bkvh0000gn/T/renovate/repos/github/barriot/r0\" -v \"/var/folders/dj/wt5h_b751qjfgf5h0fv5bkvh0000gn/T/renovate/cache\":\"/var/folders/dj/wt5h_b751qjfgf5h0fv5bkvh0000gn/T/renovate/cache\" -e PIP_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/var/folders/dj/wt5h_b751qjfgf5h0fv5bkvh0000gn/T/renovate/repos/github/barriot/r0\" ghcr.io/containerbase/sidecar:9.19.5 bash -l -c \"install-tool python 3.11.5 && install-tool poetry 1.5.1 && poetry update --lock --no-interaction python\""
DEBUG: exec completed (repository=barriot/r0, branch=renovate/python-3.x)
       "durationMs": 24799,
       "stdout": "[01:08:52.108] INFO (9): Installing tool python v3.11.5...\ninstalling v2 tool python v3.11.5\nlinking tool python v3.11.5\nPython 3.11.5\npip 23.2.1 from /opt/containerbase/tools/python/3.11.5/lib/python3.11/site-packages/pip (python 3.11)\n[01:08:58.884] INFO (9): Installed tool python in 6.7s.\ninstalling v2 tool poetry v1.5.1\n[01:08:59.353] INFO (148): Installing tool poetry v1.5.1...\nlinking tool poetry v1.5.1\nPoetry (version 1.5.1)\n[01:09:14.462] INFO (148): Installed tool poetry in 15.1s.\nUpdating dependencies\nResolving dependencies...\n\nWriting lock file\n",
       "stderr": "Creating virtualenv test-WceijsJ--py3.11 in /home/ubuntu/.cache/pypoetry/virtualenvs\n"
DEBUG: Returning updated poetry.lock (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Updated 1 package files (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Updated 1 lock files (repository=barriot/r0, branch=renovate/python-3.x)
       "updatedArtifacts": ["poetry.lock"]
DEBUG: Getting comments for #16 (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Found 0 comments (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: isBranchConflicted(main, renovate/python-3.x) (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: branch.isConflicted(): using git to calculate (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: branch.isConflicted(): false (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: 2 file(s) to commit (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Preparing files for committing to branch renovate/python-3.x (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: git commit (repository=barriot/r0, branch=renovate/python-3.x)
       "deletedFiles": [],
       "ignoredFiles": [],
       "result": {
         "author": null,
         "branch": "renovate/python-3.x",
         "commit": "69bea022fff59db4284222f7da00f5c56f1676e6",
         "root": false,
         "summary": {"changes": 2, "insertions": 3, "deletions": 3}
       }
DEBUG: Pushing refSpec renovate/python-3.x:renovate/python-3.x (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: git push (repository=barriot/r0, branch=renovate/python-3.x)
       "result": {
         "pushed": [],
         "ref": {"local": "refs/remotes/origin/renovate/python-3.x"},
         "remoteMessages": {"all": []}
       }
DEBUG: Setting current branch to main (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: latest commit (repository=barriot/r0, branch=renovate/python-3.x)
       "branchName": "main",
       "latestCommitDate": "2023-09-10T10:37:12+10:00"
 INFO: Branch updated (repository=barriot/r0, branch=renovate/python-3.x)
       "commitSha": "69bea022fff59db4284222f7da00f5c56f1676e6"
DEBUG: Ensuring PR (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: There are 0 errors and 0 warnings (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: getBranchPr(renovate/python-3.x) (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: findPr(renovate/python-3.x, undefined, open) (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Found PR #16 (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: getPrCache() (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Found existing PR (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: PR rebase requested, so skipping cache check (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: Processing existing PR (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: updatePr(16, Update python Docker tag to v3.11.5, body) (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: PR updated...prNo: 16 (repository=barriot/r0, branch=renovate/python-3.x)
 INFO: PR updated (repository=barriot/r0, branch=renovate/python-3.x)
       "pr": 16,
       "prTitle": "Update python Docker tag to v3.11.5"
DEBUG: setPrCache() (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: PR is not configured for automerge (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: setBranchCommit() (repository=barriot/r0, branch=renovate/python-3.x)
DEBUG: getBranchPr(renovate/requests-2.x) (repository=barriot/r0)
DEBUG: findPr(renovate/requests-2.x, undefined, open) (repository=barriot/r0)
DEBUG: Found PR #11 (repository=barriot/r0)
DEBUG: getPrCache() (repository=barriot/r0)
DEBUG: getBranchPr(renovate/python-3.x) (repository=barriot/r0)
DEBUG: findPr(renovate/python-3.x, undefined, open) (repository=barriot/r0)
DEBUG: Found PR #16 (repository=barriot/r0)
DEBUG: getPrCache() (repository=barriot/r0)
DEBUG: ensureDependencyDashboard() (repository=barriot/r0)
DEBUG: Closing Dependency Dashboard (repository=barriot/r0)
DEBUG: Removing any stale branches (repository=barriot/r0)
DEBUG: config.repoIsOnboarded=true (repository=barriot/r0)
DEBUG: Branch lists (repository=barriot/r0)
       "branchList": ["renovate/python-3.x", "renovate/requests-2.x"],
       "renovateBranches": ["renovate/python-3.x", "renovate/requests-2.x"]
DEBUG: remainingBranches= (repository=barriot/r0)
DEBUG: No branches to clean up (repository=barriot/r0)
DEBUG: PackageFiles.clear() - Package files deleted (repository=barriot/r0)
DEBUG: Branch summary (repository=barriot/r0)
       "cacheModified": undefined,
       "baseBranches": [{"branchName": "main", "sha": "a864255b0b4159f2592027b1cd2dd6cc78bc5163"}],
       "branches": [
         {
           "automerge": false,
           "baseBranch": "main",
           "baseBranchSha": "a864255b0b4159f2592027b1cd2dd6cc78bc5163",
           "branchName": "renovate/requests-2.x",
           "branchSha": "1105232dffa5542ed8b9734df68878eb193ffb66",
           "isModified": false,
           "isPristine": false
         },
         {
           "automerge": false,
           "baseBranch": "main",
           "baseBranchSha": "a864255b0b4159f2592027b1cd2dd6cc78bc5163",
           "branchName": "renovate/python-3.x",
           "branchSha": "69bea022fff59db4284222f7da00f5c56f1676e6",
           "isModified": false,
           "isPristine": true
         }
       ],
       "defaultBranch": "main",
       "inactiveBranches": []
DEBUG: branches info extended (repository=barriot/r0)
       "branchesInformation": [
         {
           "branchName": "renovate/requests-2.x",
           "prNo": 11,
           "prTitle": "Update dependency requests to ^2.31.0",
           "result": "done",
           "upgrades": [
             {
               "datasource": "pypi",
               "depName": "requests",
               "displayPending": "",
               "fixedVersion": "2.27.0",
               "currentVersion": "2.27.0",
               "currentValue": "^2.27.0",
               "newValue": "^2.31.0",
               "newVersion": "2.31.0",
               "packageFile": "pyproject.toml",
               "updateType": "minor",
               "packageName": "requests"
             }
           ]
         },
         {
           "branchName": "renovate/python-3.x",
           "prNo": 16,
           "prTitle": "Update python Docker tag to v3.11.5",
           "result": "done",
           "upgrades": [
             {
               "datasource": "docker",
               "depName": "python",
               "displayPending": "",
               "fixedVersion": "3.10.11",
               "currentVersion": "3.10.11",
               "currentValue": "3.10.11",
               "newValue": "3.11.5",
               "newVersion": "3.11.5",
               "packageFile": "pyproject.toml",
               "updateType": "minor",
               "packageName": "python"
             }
           ]
         }
       ]
DEBUG: Renovate repository PR statistics (repository=barriot/r0)
       "stats": {"total": 14, "open": 2, "closed": 12, "merged": 0}
DEBUG: Repository result: done, status: onboarded, enabled: true, onboarded: true (repository=barriot/r0)
DEBUG: Repository timing splits (milliseconds) (repository=barriot/r0)
       "splits": {"init": 11286, "extract": 91, "lookup": 4758, "onboarding": 1, "update": 34165},
       "total": 50330
DEBUG: Package cache statistics (repository=barriot/r0)
       "get": {"count": 19, "avgMs": 1, "medianMs": 1, "maxMs": 8},
       "set": {"count": 2, "avgMs": 79, "medianMs": 73, "maxMs": 84}
DEBUG: http statistics (repository=barriot/r0)
       "urls": {
         "https://api.github.com/graphql (POST,200)": 2,
         "https://api.github.com/repos/barriot/r0/branches/main/protection (GET,404)": 1,
         "https://api.github.com/repos/barriot/r0/issues/16/comments (GET,200)": 1,
         "https://api.github.com/repos/barriot/r0/pulls (GET,200)": 1,
         "https://api.github.com/repos/barriot/r0/pulls/16 (PATCH,200)": 1,
         "https://auth.docker.io/token (GET,200)": 1,
         "https://foo.bar/simple/requests/ (GET,-1)": 1,
         "https://index.docker.io/v2/library/python/tags/list (GET,200)": 1,
         "https://index.docker.io/v2/library/python/tags/list (GET,401)": 1
       },
       "hostStats": {
         "api.github.com": {"requestCount": 6, "requestAvgMs": 791, "queueAvgMs": 0},
         "auth.docker.io": {"requestCount": 1, "requestAvgMs": 794, "queueAvgMs": 0},
         "foo.bar": {"requestCount": 1, "requestAvgMs": 2, "queueAvgMs": 0},
         "index.docker.io": {"requestCount": 2, "requestAvgMs": 1861, "queueAvgMs": 1}
       },
       "totalRequests": 10
DEBUG: Package lookup durations (repository=barriot/r0)
       "pypi": {"count": 1, "averageMs": 3199, "totalMs": 3199, "maximumMs": 3199},
       "docker": {"count": 1, "averageMs": 4712, "totalMs": 4712, "maximumMs": 4712}
DEBUG: dns cache (repository=barriot/r0)
       "hosts": []
 INFO: Repository finished (repository=barriot/r0)
       "cloned": true,
       "durationMs": 50330
DEBUG: Checking file package cache for expired items
DEBUG: Deleted 0 of 188 file cached entries in 73ms
DEBUG: Renovate exiting
  ```
</details>

**Review by rarkins — APPROVED** [review]

File: lib/modules/manager/poetry/extract.ts [hunk 1]
```diff
@@ -37,7 +38,12 @@ export async function extractPackageFile(
       if (dep.currentValue) {
         pythonVersion = dep.currentValue;
       }
-      return null;
+      return {
+        ...dep,
+        datasource: DockerDatasource.id,
```
Comment (viceice) [inline lib/modules/manager/poetry/extract.ts:43]: we should use our containerbase python as source. there are a lot docker tags which can cause issues with poetry versioning.

Comment (zeshuaro) [reply lib/modules/manager/poetry/extract.ts:43]: Is that already available for renovate or it needs to be introduced?

Comment (viceice) [reply lib/modules/manager/poetry/extract.ts:43]: it's available.

https://github.com/containerbase/python-prebuild

Comment (zeshuaro) [reply lib/modules/manager/poetry/extract.ts:43]: Raised https://github.com/renovatebot/renovate/pull/24463

**Pull request merged**